src/raftstore/store/peer.rs

```diff
@@ -802,6 +813,83 @@ impl Peer {
         cmd.call(resp);
     }
 
+    fn is_uptodate(&self, matched_index: u64) -> bool {
+        matched_index >= self.get_store().truncated_index()
+    }
+
+    // Calculate the count of uptodate peers, and whether the specified peer is uptodate.
+    fn calc_uptodate(&self, peer_id: u64, progress: &HashMap<u64, Progress>) -> (usize, bool) {
+        let mut peer_included = false;
+        let mut uptodate = 0;
+        for (id, pr) in progress {
+            if self.is_uptodate(pr.matched) {
+                uptodate += 1;
+                if *id == peer_id {
+                    peer_included = true;
+                }
+            }
+        }
+        (uptodate, peer_included)
+    }
+
+    /// Check whether it's safe to propose the specified conf change request.
+    /// It's safe iff at least the quorum of the Raft group is still up to date
+    /// right after that conf change is applied.
+    /// Define the total number of nodes in current Raft cluster to be `total`.
+    /// To ensure the above safety, if the cmd is
+    /// 1. A `AddNode` request
+    ///    Then at least '(total + 1)/2 + 1' nodes need to be up to date for now.
+    /// 2. A `RemoveNode` request
+    ///    Then at least '(total - 1)/2 + 1' other nodes (the node about to be removed is excluded)
+    ///    need to be up to date for now.
+    fn check_conf_change(&self, cmd: &RaftCmdRequest) -> Result<()> {
```

Comment:
A more clear way to do this check is apply the conf change and then check the quorum.